core/src/display_object/movie_clip.rs

```diff
@@ -550,6 +578,132 @@ impl<'gc> MovieClip<'gc> {
 
         actions.into_iter()
     }
+
+    pub fn set_fill_style(
+        self,
+        context: &mut UpdateContext<'_, 'gc, '_>,
+        style: Option<FillStyle>,
+    ) {
+        let mut mc = self.0.write(context.gc_context);
+
+        // TODO: If current_fill is not closed, we should close it and also close current_line
+
+        if let Some(existing) = mc.current_fill.take() {
+            mc.custom_fills.push(existing);
+        }
+        if let Some(style) = style {
+            mc.current_fill = Some((style, Vec::new()));
+        }
+
+        mc.dirty_shape = true;
+    }
+
+    pub fn clear(self, context: &mut UpdateContext<'_, 'gc, '_>) {
+        let mut mc = self.0.write(context.gc_context);
+        mc.current_fill = None;
+        mc.current_line = None;
+        mc.custom_fills.clear();
+        mc.custom_lines.clear();
+        mc.custom_edge_bounds = BoundingBox::default();
+        mc.custom_shape_bounds = BoundingBox::default();
+        mc.dirty_shape = true;
+    }
+
+    pub fn set_line_style(
+        self,
+        context: &mut UpdateContext<'_, 'gc, '_>,
+        style: Option<LineStyle>,
+    ) {
+        let mut mc = self.0.write(context.gc_context);
+
+        if let Some(existing) = mc.current_line.take() {
+            mc.custom_lines.push(existing);
+        }
+        if let Some(style) = style {
+            mc.current_line = Some((style, Vec::new()));
+        }
+
+        mc.dirty_shape = true;
+    }
+
+    pub fn draw_command(self, context: &mut UpdateContext<'_, 'gc, '_>, command: DrawCommand) {
+        let mut mc = self.0.write(context.gc_context);
+
+        let mut include_last = false;
+
+        match command {
+            DrawCommand::MoveTo { .. } => {}
+            DrawCommand::LineTo { x, y } => {
+                mc.custom_shape_bounds.encompass(x, y);
+                mc.custom_edge_bounds.encompass(x, y);
+                include_last = true;
+            }
+            DrawCommand::CurveTo { x1, y1, x2, y2 } => {
+                mc.custom_shape_bounds.encompass(x1, y1);
+                mc.custom_shape_bounds.encompass(x2, y2);
+                mc.custom_edge_bounds.encompass(x1, y1);
+                mc.custom_edge_bounds.encompass(x2, y2);
+                include_last = true;
+            }
+        }
+
+        if let Some((_, commands)) = &mut mc.current_line {
+            commands.push(command.clone());
+        }
+        if let Some((_, commands)) = &mut mc.current_fill {
+            commands.push(command);
+        }
+
+        if include_last {
+            if let Some(command) = mc
+                .current_fill
+                .as_ref()
+                .and_then(|(_, commands)| commands.last().cloned())
```

Comment:
Don't think we need `cloned()` here, can match on `&DrawCommand::MoveTo` etc.

Reply:
We'd need to turn all the uses of x/y into `*x` `*y`, which I find less readable. A copy of the complete object happens one way or another, I think this way is preferable.

Reply:
If I match on `&DrawCommand::MoveTo { x, y }` with the `&`, it's smart enough to match the parameters by value

Reply:
Added a commit that does this (but by dereferencing the match entirely, per clippy suggestions)